Chepulis.mikhail

MChepulis opened the pull request.

MChepulis committed:
add-lab_1

add build script
add lab_1

MChepulis committed:
Update lab_1.cpp

add same path check
add frequency positive (>=0) check
fix braces spaces

MChepulis committed:
Update lab_1.cpp

Перенес константы из деректив препроцессора в глобальные переменные

справил флаг LOG_USER на флаг LOG_ERR там, где в лог выводится информация об ошибке

исправил SIGTERM на EXIT_SECCESS

перенёс ветвление из  main  в Init

добавил функцию удаления pid-файла и её вызов при завершении работы с ошибкой

заменил NULL на nullptr

исправил отступ

исправил ошибку в названии функции "CompareExtantion"

MChepulis committed:
Update lab_1.cpp

rename extantion to extantion

MChepulis committed:
Update lab_1.cpp

проверил в словаре, правильное правописание "exstension"

MChepulis committed:
remove is_need_work  flag

remove is_need_work  flag

MChepulis committed:
remove is_need_work

удалил остатки переменной is_need_work

MChepulis committed:
Add lab_2

add lab_2:
add mq connection
add pipe connection
add shm_mem connection
add host

MChepulis committed:
remove bits/types/siginfo.h

remove <bits/types/siginfo.h>
add <signal.h>

MChepulis committed:
delete copy C-TOR

delete copy C-TOR for singletone classes

MChepulis committed:
Fix remarks of lab2

Подправил отступы
ClientHandler теперь не синглтон
теперь сервер создаёт потоки для общения с клиентами, а не процессы
Дабавил проверку на реальность даты
Убрал неиспользуемый семафор

MChepulis committed:
Fix remarks2 of lab2

убрал ненужный вывод в лог файл
разбил функцию Server::Start() на 4 подфункций
1)ReadClientCount() - считывает с клавиатуры кол-во клиентов
2) void CreateClient(ClientHandler *handler)
3) void StartHandlerActivity(ClientHandler *handler)
4) void Server::Run()

добавил адекватную реакцию Сервера на то, что какой-то из( потоков взаимодействия)/клиентов не смог открыться

MChepulis committed:
Fix remarks3 in lab2

перенёс все "сторонние функции" в приватные методы класса
убрал функцию ToString, заменив её на std::to_string

MChepulis committed:
add-Lab_3

lab_3 init commit

KotsMikhail reviewed (changes requested):
Не собирается, сообщение в консоли следующее:
g++   -Wall -std=c++11 -I ./src/ ./src/*.cpp ./src/*.h  -o Lab_3(stack) -lpthread -latomic
/bin/sh: 1: Syntax error: "(" unexpected
makefile:10: recipe for target 'build_all' failed
make: *** [build_all] Error 2
Под какую версию make вы писали свой скрипт сборки?

MChepulis committed:
fix makefile

исправил флаги
исправил название итогового файла

MChepulis committed:
add cammand line argument

1) N_big  - размер массива для единоразового теста
2) R_num - кол-во писателей для единоразового теста
3) W_num - кол-во читателей для единоразового теста
4) N -  размер массива
5) max_thread - макс кол-во потоков для перебора

Убрал вывод сообщения о том, что превышен лимит ожидания при опбращению к стеку

KotsMikhail commented on chepulis.mikhail/lab_3/src/CommonTester.cpp:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    bool result = true;
+    pthread_t *R_threads = (pthread_t *) malloc((num_of_reader) * sizeof(pthread_t));
```
Память в с++ выделяется не так

MChepulis replied:
исправил

KotsMikhail replied:
Не вижу коммит

MChepulis replied:
исправлено в коммите "Fix remarks of lab3" (093a35f)

KotsMikhail commented on chepulis.mikhail/lab_3/src/main.cpp:
```diff
@@ -0,0 +1,280 @@
+
+    IStack *my_block_stack = new BlockedStack;
+    IStack *my_non_block_stack = new NonBlockedStack;
+    bool is_need_test[4];
```
Зачем нужна эта переменная?

MChepulis replied:
эта переменная позволяет во время отладки быстро включать/отключать необходимые тесты

KotsMikhail replied:
В релизной версии программы отладочного кода быть не должно. А ещё вам будет сильно проще, если у вас будет абстрактный класс для тестов и вам достаточно будет 1 раз перебрать все тесты и структуры, которые вы хотите

KotsMikhail commented on chepulis.mikhail/lab_3/src/NonBlockedStack.h:
```diff
@@ -0,0 +1,56 @@
+        CountedNodePtr() : external_count(0), ptr(nullptr) {}
+
+        CountedNodePtr(Node *ptr_) : external_count(0), ptr(ptr_) {}
+    }__attribute__((__aligned__(16)));
```
Зачем вам здесь выравнивание?

MChepulis replied:
т.к. атомарные функции требуют выравнивания для атомарного типа

KotsMikhail reviewed (changes requested):
Не очень понятно, что программа принимает на вход, и что выводит. Огромный вывод каждого запуска эксперимента выводить не обязательно, достаточно среднего значения

MChepulis committed:
Fix remarks of lab3

исправил выделения памяти под массивы
исправил освобожедния памяти
исправил тело деструктора BlockedStack

MChepulis commented:
Вход - 5 параметров: (
3 - для единоразового запуска, например посмотреть работ у с конкретным числом потоков для массива большой размерности 
2 - для переборного теста
)
1) размер массива для единоразового теста
2) кол-во писателей для единоразового теста
3) кол-во читателей для единоразового теста
4) размер массива для переборного теста
5) макс кол-во потоков для перебора
Пример: 100000 10 10 1000 10
Вывод:
делается один вывод для каждой пары (кол-во писателей, кол-во читателей), время усредняется по 10 повторениям

KotsMikhail commented on chepulis.mikhail/lab_3/src/CommonTester.h:
```diff
@@ -0,0 +1,101 @@
+
+    int *array;
+
+    struct thread_W_args {
```
Эти поля стоит сделать полями CommonTester и сделать методы Read и Write не статическими, тогда поля будут доступны в этих функциях, а в потоки можно передавать указатель на метод и не тащить за ним структуру

MChepulis replied:
Удалил переменную is_need_test

MChepulis replied:
По поводу методов Read и Write:
для вызова функции pthread_create требуется статическая функция аргументы которой можно передать по void*
если эти функции перестанут быть статическими, то я не смогу передавать их  в pthread_create. Делать класс тестера синглтоном (или какие-то иные обходные пути) я не хочу, т.к. этим я накладываю ограничение на класс-тестер, хотя могу просто передать аргументы как void*

Собрал всех тестировщиков в один общий класс

Очистил main (перенёс все функции тестирования в отдельный класс)

Убрал выбрасывание исключений при превышении времени обращения к стеку

SWAGLORD1 replied:
Почему вы уверены, что start_routine для pthread должна быть статической? В документации об этом ни слова.

KotsMikhail reviewed (changes requested):
Для теста читателей и писателей тесты с разными комбинациями потоков делать не нужно. когда у вас один тип потоков, эти тесты фундаментально ничем не отличаются
И перебор потоков с замером времени следует делать внутри классов для тестов, а не в main
Также, ожидание по таймеру в обоих структурах стоит убрать, иначе в случае огромного кол-ва потоков ваша программа падает

MChepulis committed:
remove is_need_test

удалил переменную is_need_test

MChepulis committed:
fix remarks2 for lab_3

перенёс функции тестирования из mein  в отдельный класс
убрал все выбрасывания исключений по таймауту времени ожидания
собрал всех тестировшиков в один единый класс

KotsMikhail reviewed (changes requested):
Ещё раз: для теста читателей и писателей тесты с разными комбинациями потоков делать не нужно. когда у вас один тип потоков, эти тесты фундаментально ничем не отличаются
Не нужно выводить результаты отдельно для каждого сочетания читателей/писателей в общем тесте, просто выведите среднее для всех прогонов

MChepulis committed:
fix remarks3 for lab_3

изменил формат вывода
для теста с перебором теперь выводится интервал значений кол-ва потоков и среднее время работы

MChepulis commented:
Теперь тест читателей и писателей проводится всегда только для максимального кол-ва потоков
Изменил формат вывода
Теперь тип тестируемого стека выводится на одной строчке с данными теста
Для переборного теста выводится только одна запись: интервал перебранного кол-ва потоков, среднее время работы, общий результат тестов

MChepulis committed:
decorated output

немного приукрасыл вывод

SWAGLORD1 commented:
Засчитано
